src/server/scene/basic_idle_hub.cpp

```diff
@@ -0,0 +1,246 @@
+
+    void idle() override
+    {
+        for_each_observer(&IdleStateObserver::idle);
+    }
+
+    void active() override
+    {
+        for_each_observer(&IdleStateObserver::active);
+    }
+};
+
+ms::BasicIdleHub::BasicIdleHub(
+    std::shared_ptr<time::Clock> const& clock,
+    mt::AlarmFactory& alarm_factory)
+    : clock{clock},
+      alarm{alarm_factory.create_alarm(
+          std::make_unique<AlarmCallback>(mutex, [this](std::unique_lock<std::mutex>& lock)
+            {
+                alarm_fired(lock);
+            }))}
+{
+    poke();
+}
+
+ms::BasicIdleHub::~BasicIdleHub()
+{
+    alarm->cancel();
+}
+
+void ms::BasicIdleHub::poke()
+{
+    std::unique_lock<std::mutex> lock{mutex};
+    poke_time = clock->now();
+    schedule_alarm(lock, poke_time);
+    if (!idle_multiplexers.empty())
+    {
+        auto const idle = std::move(idle_multiplexers);
+        idle_multiplexers.clear();
+        lock.unlock();
+        for (auto const& multiplexer : idle)
+        {
+            multiplexer->active();
+        }
+    }
+}
+
+void ms::BasicIdleHub::register_interest(
+    std::weak_ptr<IdleStateObserver> const& observer,
+    time::Duration timeout)
+{
+    register_interest(observer, system_executor, timeout);
+}
+
+void ms::BasicIdleHub::register_interest(
+    std::weak_ptr<IdleStateObserver> const& observer,
+    Executor& executor,
+    time::Duration timeout)
+{
+    std::unique_lock<std::mutex> lock{mutex};
+    auto const iter = timeouts.find(timeout);
+    bool is_active = true;
+    if (iter == timeouts.end())
+    {
+        auto const multiplexer = std::make_shared<Multiplexer>();
+        multiplexer->register_interest(observer, executor);
+        timeouts.insert({timeout, multiplexer});
+        first_timeout = timeouts.begin()->first;
+        auto const current_time = clock->now() - poke_time;
+        if (alarm_timeout && alarm_timeout.value() <= timeout)
+        {
+            // The alarm will be fired before we hit our timeout
+            is_active = true;
+        }
+        else if (current_time < timeout)
+        {
+            // The alarm will be fired after we hit our timout,
+            // but we have not hit it yet so the alarm needs to be moved up
+            is_active = true;
+            alarm_timeout = timeout;
+            alarm->reschedule_in(std::chrono::duration_cast<std::chrono::milliseconds>(timeout - current_time));
+        }
+        else
+        {
+            // This timeout has already been passed
+            is_active = false;
+            idle_multiplexers.push_back(multiplexer);
+        }
+    }
+    else
+    {
+        iter->second->register_interest(observer, executor);
+        is_active = (alarm_timeout && alarm_timeout.value() <= timeout);
+    }
+    lock.unlock();
+    if (auto const shared = observer.lock())
+    {
+        if (is_active)
+        {
+            shared->active();
+        }
+        else
+        {
+            shared->idle();
+        }
+    }
```

Comment:
I feel like there's a race here, where a timeout is registered just as it would timeout
|Thread A | Timer Thread|
|-|-|
|`alarm->reschedule_in(1)`||
|`is_active = true`||
| `lock.unlock()` ||
| <Thread loses timeslice, and is scheduled out> | |
| | <Timer fires> |
| | <Calls `observer->idle()`> |
| `if (is_active)` branch is taken| |
| `observer->active()` is called | |

This results in the observer believing Mir is currently active, and once `poke()` is called will result in `active()` being called a second time in a row.

On the one hand, this is not actually likely to happen (with the current code), as the observers are registered once at startup. On the other hand, *debugging* this should it occur would be a nightmare; it would manifest as *sometimes* it appearing that an idle timeout is failing to fire.

Reply:
A corollary of this is that code *using* the `IdleHub` almost certainly needs to be using a serialising `Executor` if it wants to act on *state*, rather than just transitions.

Reply:
Just an observation at this time - I've not understood this code well enough to decide if there are real problems.

This discussion is reminiscent of the early "observer" iterations that led to `BasicObservers`. The story is here:

https://accu.org/journals/overload/22/124/griffiths_2040/